File: tokenizer/stringcreator.py
import os
import logging
import time
from concurrent.futures import ProcessPoolExecutor, as_completed
from pdfminer.high_level import extract_text

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def extract_text_from_pdfs(pdf_file_paths):
    """Extracts text from a list of PDF files using pdfminer."""
    texts = []
    for pdf_file_path in pdf_file_paths:
        try:
            text = extract_text(pdf_file_path)
            if not text.strip():
                logger.warning(f"No text found in {pdf_file_path}")
            texts.append((os.path.splitext(os.path.basename(pdf_file_path))[0], text))
        except Exception as e:
            logger.error(f"An unexpected error occurred while reading {pdf_file_path}: {e}")
    return texts

def save_text_to_file(text_data, output_directory):
    """Saves the extracted text to individual files."""
    for filename, text in text_data:
        output_file_path = os.path.join(output_directory, f"{filename}.txt")
        try:
            with open(output_file_path, "w", encoding="utf-8") as file:
                file.write(text)
            logger.info(f"Text saved to {output_file_path}")
        except Exception as e:
            logger.error(f"An error occurred while saving the text to file {output_file_path}: {e}")

def main():
    start_time = time.time()
    directory_path = r"C:\Users\rafa0\Desktop\pj\laplace\LAPLACE\data\book"
    output_directory = os.path.join(os.getcwd(), "saved_files")
    os.makedirs(output_directory, exist_ok=True)

    # Check if the directory is accessible
    if not os.access(directory_path, os.R_OK):
        logger.error(f"Cannot read from directory: {directory_path}")
        exit(1)

    # Extract text from PDFs
    pdf_file_paths = [os.path.join(directory_path, filename) for filename in os.listdir(directory_path) if filename.endswith(".pdf")]
    batch_size = 4  # Adjust the batch size as needed
    with ProcessPoolExecutor() as executor:
        futures = []
        for i in range(0, len(pdf_file_paths), batch_size):
            batch = pdf_file_paths[i:i+batch_size]
            futures.append(executor.submit(extract_text_from_pdfs, batch))

        text_data = []
        for future in as_completed(futures):
            try:
                text_data.extend(future.result())
            except Exception as e:
                logger.error(f"An error occurred while processing a batch of PDF files: {e}")

    # Save text to individual files
    save_text_to_file(text_data, output_directory)

    end_time = time.time()
    execution_time = end_time - start_time
    logger.info(f"Execution time: {execution_time} seconds")

if __name__ == '__main__':
    main()

File: tokenizer/string_creator.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <future>
#include <podofo/podofo.h>

std::string extractTextFromPDF(const std::string& pdfFilePath) {
    std::string text;
    try {
        PoDoFo::PdfMemDocument pdfDocument(pdfFilePath.c_str());
        for (int pageNum = 0; pageNum < pdfDocument.GetPages(); ++pageNum) {
            PoDoFo::PdfRefCountedBuffer buffer;
            PoDoFo::PdfSimpleWriter writer(&buffer);
            pdfDocument.WriteCurrentPage(&writer, pageNum);
            std::string pageText = buffer.GetBuffer();
            text += pageText;
        }
    } catch (const PoDoFo::PdfError& e) {
        std::cerr << "Error reading PDF file " << pdfFilePath << ": " << e.what() << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Unexpected error reading " << pdfFilePath << ": " << e.what() << std::endl;
    }
    return text;
}

void saveTextToFile(const std::string& text, const std::string& outputFilePath) {
    try {
        std::ofstream outputFile(outputFilePath, std::ios::out | std::ios::trunc);
        outputFile << text;
        std::cout << "Text saved to " << outputFilePath << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error saving text to file: " << e.what() << std::endl;
    }
}

int main() {
    std::string directoryPath = "C:\\Users\\rafa0\\Desktop\\pj\\laplace\\LAPLACE\\data\\books";
    std::string outputFilePath = "output.txt";

    std::vector<std::future<std::string>> futures;
    for (const auto& entry : std::filesystem::directory_iterator(directoryPath)) {
        if (entry.path().extension() == ".pdf") {
            futures.emplace_back(std::async(std::launch::async, extractTextFromPDF, entry.path().string()));
        }
    }

    std::string textData;
    for (auto&& future : futures) {
        textData += future.get();
    }

    saveTextToFile(textData, outputFilePath);

    return 0;
}